lib/mapping-transform.js

```diff
@@ -1,34 +1,27 @@
 const {Transform} = require('stream');
 const RiffError = require('./riff-error');
 
+/*
+ * Wraps promise-based (using async or raw API) and synchronous
+ * functions to pipe-able stream.
+ */
 module.exports = class MappingTransform extends Transform {
 
     constructor(fn, options) {
         super(options);
         this._function = fn;
-
     }
 
     _transform(chunk, _, callback) {
-        try {
-            Promise.resolve(this._function(chunk))
-                .then(
-                    this._handleResult(callback),
-                    this._handleError.bind(this)
-                );
-        } catch (err) {
-            Promise.resolve(err).then(this._handleError.bind(this));
-        }
-    }
-
-
-    _handleResult(callback) {
-        return (result) => {
-            callback(null, result);
-        }
-    }
-
-    _handleError(err) {
-        this.emit('error', new RiffError('request-reply-function-runtime-error', err));
+        // the function may throw, in which case the error is caught upstream by the input unmarshaller
```

Comment:
I think error handling throughout the pipeline (inc. unmarshallers and marshallers) needs to be revisited. It's a hot mess right now and is very hard to comprehend.

Reply:
we have three user definable functions where throwing is legitimate:
- the function
- $init
- $destroy

If the user defines these as async functions, they will never throw, but we can't force that.

Reply:
are you suggesting in that particular case to catch the potential error of the `this._function(chunk)` call?

Reply:
Up :) Can you elaborate on what should be done concretely here @scothis? I am not sure I understand.

Reply:
`this._function` is the user's function. So it can throw, we need to handle that case elegantly by passing the error to the callback. Currently the _transform method would propagate the error thrown from user code.

Many promise libraries have a `try` method which is nice, but that's not part of the standard spec. I'd do something like:

```js
Promise.resolve()
    .then(() => this.function(chunk))
    .then(result => callback(null, result))
    .catch(err => callback(new RiffError('request-reply-function-runtime-error', err)))
```